src/handlers/bbb/api.js

```diff
@@ -0,0 +1,122 @@
+        });
+}
+
+const join_translations = {
+    de: {
+        not_found: 'Der gewünschte Raum existiert nicht.',
+        server_error:
+            'Beim Beitreten ist leider ein unerwarter Serverfehler aufgetreten. Bitte versuche es später erneut.',
+        admin_only: 'Du darfst diesem Raum nicht beitreten.',
+    },
+    en: {
+        not_found: 'The requested room does not exist.',
+        server_error:
+            'Unfortunately, an unexpected error occurred while trying to join the requested room. Please try again later.',
+        admin_only: 'You cannot join this room.',
+    },
+};
+async function joinRoom(request, h) {
+    const is_admin = request.state.bbb_admin_token === config.admin_token;
+
+    const { lang, slug } = request.params;
+
+    const meeting = await request.server.methods.knex('bbb').select().where({ slug }).first();
+    if (!meeting) return h.view('render-text', { text: join_translations[lang].not_found }).code(404);
+
+    if (!is_admin && meeting.admin_only)
+        return h.view('render-text', { text: join_translations[lang].admin_only }).code(404);
+
+    const params = new URLSearchParams(meeting.params);
+
+    const name = request.query.name;
+    if (!name) return h.view(`bbb-enter-name.${lang}.html`, { room_name: params.get('name') });
+
+    const is_meeting_running = await bbb('isMeetingRunning', { meetingID: slug }).then((r) => r.running);
+
+    if (!is_meeting_running && !is_admin && !meeting.anonymous_users_can_start)
+        return h.view(`bbb-wait-for-start.${lang}.html`, { room_name: params.get('name') }).header('refresh', '60');
+
+    const bbb_meeting = await bbb('getMeetingInfo', { meetingID: slug })
+        .catch(() => bbb('create', meeting.params))
+        .catch((err) => {
+            console.error('Failed to create BBB meeting:', err);
+            return h.view('render-text', { text: join_translations[lang].server_error }).code(500);
+        });
+
+    const join_url = await bbb('join', {
+        fullName: name,
+        meetingID: bbb_meeting.meetingID,
+        // This is quite silly. These passwords actually serve no security purpose. Knowing them gets you nothing since
+        // you need to know the API key to generate the checksum ("signature").
+        //
+        // The BBB devs appear to have realized this as well since according to the API docs
+        // (https://docs.bigbluebutton.org/development/api/#get-join), the `password` parameter is deprecated when
+        // joining and you're instead just supposed to use `role=MODERATOR` or `role=VIEWER`. However, for some reason
+        // this doesn't appear to actually work (might be caused by Senfcall's proxy?).
+        //
+        // But, since we need to fetch the meeting details from the API anyway, I'm not "emulating" the `role` parameter
+        // by hardcoding the passwords.
+        password: is_admin ? bbb_meeting.moderatorPW : bbb_meeting.attendeePW,
+        createTime: bbb_meeting.createTime,
```

Comment:
This adds no extra security benefit, the way you implemented it now.

Reply:
Why not? According to the [docs](https://docs.bigbluebutton.org/development/api/#get-join):

> This prevents a user from reusing their join URL for a subsequent session with the same meetingID.

Maybe I'm missing something but my understanding is that this meant to prevent users from bypassing the checks implemented in the frontend, e.g. in our frontend:

* I create a conference `test123` that anyone can join.
* Someone joins the conference, it is thus created through the BBB API and the `createTime` is set (let's say it's `1`).
* User U, not an admin, joins the conference and sneakily stores their join URL.
* The conference is ended and deleted.
* Later on, I create a new conference `test123`, but this time, I only allow admins to join.
* I join the conference, thus there's a new `createTime` (let's say `2`).
* U is not supposed to be able to join this conference since they are not an admin. Our frontend correctly blocks their join request. But U, channeling their best hackerperson energy, tries to use their old join link to join. This also fails since `2 != 1`.